# Week 6 Exercises
### Download the State-level baby names data from https://www.ssa.gov/oact/babynames/limits.html

In [1]:
%pylab inline
import seaborn as sns
import os
import glob
import pandas as pd

Populating the interactive namespace from numpy and matplotlib


## Exercise 1

* Construct a DataFrame with baby names data coming from the namesbystate.zip file. The DataFrame columns should have names “state”, “sex”, “year”, “name”, and “count”. You don’t need to save this data to a single csv file; if you do, the file size will be about 130 MB.

* Check: The DataFrame should have 6,215,834 rows.

In [2]:
file_names = glob.glob("/Users/victorhuang/Documents/MTH448/namesbystate/*.TXT")
frames = [pd.read_csv(f, names=["state", "sex", "year", "name", "count"]) for f in file_names]
frames
df = pd.concat(frames)
df

state sex  year     name  count
0        IN   F  1910     Mary    619
1        IN   F  1910    Helen    324
2        IN   F  1910     Ruth    238
3        IN   F  1910  Dorothy    215
4        IN   F  1910  Mildred    200
...     ...  ..   ...      ...    ...
33701    DE   M  2021   Thiago      5
33702    DE   M  2021   Travis      5
33703    DE   M  2021     Troy      5
33704    DE   M  2021   Walker      5
33705    DE   M  2021     Zayn      5

[6311504 rows x 5 columns]

## Exercise 2
* Compute a DataFrame that lists the total number of babies recorded each year.

* Check: There were 201,484 babies recorded in 1880 and 3,305,259 in 2020.

In [3]:
file_names = glob.glob("/Users/victorhuang/Documents/MTH448/names/yob*.txt")
file_names
years = [int(f[-8:-4]) for f in file_names]
frames = [pd.read_csv(f, names=["name", "sex", "count"]) for f in file_names]

df = pd.concat(frames,keys = years)
df = df.reset_index(level=0)
df = df.rename({"level_0": "year"}, axis=1)
df = df.sort_values(by="year")
df

year      name sex  count
1761   1880   Asberry   M      6
1341   1880      Vern   M     19
1340   1880   Stewart   M     19
1339   1880  Randolph   M     19
1338   1880    Lucien   M     19
...     ...       ...  ..    ...
21029  2021    Dansby   M     32
21030  2021    Darvin   M     32
21031  2021    Dayson   M     32
21033  2021       Dio   M     32
31112  2021      Riko   M      5

[2052781 rows x 4 columns]

In [4]:
df.groupby("year")["count"].sum().reset_index()

year    count
0    1880   201484
1    1881   192690
2    1882   221533
3    1883   216944
4    1884   243461
..    ...      ...
137  2017  3570234
138  2018  3508655
139  2019  3461444
140  2020  3327419
141  2021  3361375

[142 rows x 2 columns]

## Exercise 3
* Compute a DataFrame that lists the number of male babies named “John” for each year.

* Check: There were 9,655 such babies recorded in 1880 and 8,180 in 2020.



In [5]:
df_m_john = df[(df["name"]== "John")&(df["sex"]=="M")]
df_m_john.head()

year  name sex  count
942   1880  John   M   9655
938   1881  John   M   8769
1028  1882  John   M   9557
1054  1883  John   M   8894
1172  1884  John   M   9388

In [6]:
# Check
df_m_john.loc[(df_m_john["year"] == 1880) | (df_m_john["year"] == 2020) ]

year  name sex  count
942    1880  John   M   9655
17473  2020  John   M   8236

## Exercise 4

* Compute a DataFrame that lists how many different names were used each year for males and how many for females.

* Check: In 1880 there were 942 different names used for females and 1,058 for males. In 2020 these numbers were 17,360 for females and 13,911 for males.

In [7]:
df_four = df.groupby(["sex","year"])["name"].nunique().reset_index()

In [8]:
df_four.loc[(df_four["year"]==1880) | (df_four["year"] == 2020)]

sex  year   name
0     F  1880    942
140   F  2020  17447
142   M  1880   1058
282   M  2020  14006

## Exercise 5
* Compute a DataFrame that for each name shows in which year the name appeared in the records for the first time.

* Check: Here are the first recorded years for a few names: Aaban 2007, Aabha 2011, Aabid 2003, Aabidah 2018.

In [9]:
print('Aaban',df[df["name"]=='Aaban']["year"].min())
print('Aabha',df[df["name"]=='Aabha']["year"].min())
print('Aabid',df[df["name"]=='Aabid']["year"].min())
print('Aabidah',df[df["name"]=='Aabidah']["year"].min())

Aaban 2007
Aabha 2011
Aabid 2003
Aabidah 2018


# Exercise 6
* Compute a DataFrame that shows what was the most popular name for males and the most popular name for females each year.

* Check: The most popular names in 1880 were John and Mary, and in 2020 Liam and Olivia.

In [10]:
df_six = df.groupby(['year', 'sex', 'name'])['count'].max().groupby(["year","sex"]).idxmax().reset_index()
df_six     


year sex              count
0    1880   F    (1880, F, Mary)
1    1880   M    (1880, M, John)
2    1881   F    (1881, F, Mary)
3    1881   M    (1881, M, John)
4    1882   F    (1882, F, Mary)
..    ...  ..                ...
279  2019   M    (2019, M, Liam)
280  2020   F  (2020, F, Olivia)
281  2020   M    (2020, M, Liam)
282  2021   F  (2021, F, Olivia)
283  2021   M    (2021, M, Liam)

[284 rows x 3 columns]